pkg/controllermanager/controller/controllerinstallation/controllerinstallation_control.go

```diff
@@ -384,6 +409,52 @@ func (c *defaultControllerInstallationControl) isResponsible(controllerInstallat
 	return false, nil
 }
 
+func (c *defaultControllerInstallationControl) cleanOldExtensionKinds(ctx context.Context, seedClient client.Client, controllerRegistration *gardencorev1alpha1.ControllerRegistration) (bool, error) {
+	var fns []flow.TaskFn
+
+	extensionGroupVerions := kubernetes.SeedScheme.PrioritizedVersionsForGroup(gardenextensions.GroupName)
+	for _, res := range controllerRegistration.Spec.Resources {
+		for _, gv := range extensionGroupVerions {
+			var (
+				objList = &unstructured.UnstructuredList{}
+				gvk     = gv.WithKind(res.Kind)
+			)
+			objList.SetGroupVersionKind(gvk)
+
+			if err := seedClient.List(ctx, nil, objList); err != nil {
+				return false, err
+			}
+
+			if err := meta.EachListItem(objList, func(obj runtime.Object) error {
+				delFunc := func(ctx context.Context) error {
+					return seedClient.Delete(ctx, obj)
+				}
+				fns = append(fns, delFunc)
+				return nil
+			}); err != nil {
+				return false, err
+			}
+		}
+	}
+	errs := flow.Parallel(fns...)(ctx)
+	multiErrs, ok := errs.(*multierror.Error)
+	if !ok {
+		return true, errs
+	}
+
+	var result error
+	for _, err := range multiErrs.WrappedErrors() {
+		if !apierrors.IsNotFound(err) {
+			result = multierror.Append(result, err)
+		}
+	}
+
+	if result != nil {
```

Comment:
```suggestion
	return result != nil, result
```

Not sure, though, whether that's really better readable :-D